File: lib/ex_gather_web/packets/packet_writer.ex
defmodule ExGatherWeb.PacketWriter do
  @spec build() :: binary()
  def build do
    <<>>
  end

  @spec int64(binary(), integer()) :: binary()
  def int64(packet, value) do
    <<packet::bytes, value::little-signed-integer-size(64)>>
  end

  @spec uint64(binary(), integer()) :: binary()
  def uint64(packet, value) do
    <<packet::bytes, value::little-unsigned-integer-size(64)>>
  end

  @spec uint16(binary(), integer()) :: binary()
  def uint16(packet, value) do
    <<packet::bytes, value::little-unsigned-integer-size(16)>>
  end

  @spec uint8(binary(), integer()) :: binary()
  def uint8(packet, value) do
    <<packet::bytes, value::little-unsigned-integer-size(8)>>
  end

  @spec int32(binary(), integer()) :: binary()
  def int32(packet, value) do
    <<packet::bytes, value::little-integer-size(32)>>
  end

  @spec bool(binary(), boolean()) :: binary()
  def bool(packet, true), do: packet <> <<1>>
  def bool(packet, false), do: packet <> <<0>>

  @spec string(binary(), String.t()) :: binary()
  def string(packet, value) do
    value = if value, do: value, else: ""
    <<packet::bytes, byte_size(value)::little-unsigned-integer-size(32), value::binary>>
  end

  def append(packet, binary) do
    <<packet::bytes, binary::bytes>>
  end

  def reduce(packet, enum, fun) do
    Enum.reduce(enum, packet, fun)
  end
end

File: assets/js/packets/packet_writer.js
export default class PacketWriter {
  constructor() {
    this.buffer = new ArrayBuffer(0);
    this.view = new DataView(this.buffer);
    this.length = 0;
  }

  // Returns the final binary buffer
  build() {
    return this.buffer.slice(0, this.length);
  }

  // Appends an 8-bit unsigned integer
  uint8(value) {
    this.ensureCapacity(1);
    this.view.setUint8(this.length, value);
    this.length += 1;
    return this;
  }

  // Appends an 16-bit unsigned integer
  uint16(value) {
    this.ensureCapacity(2);
    this.view.setUint16(this.length, value);
    this.length += 2;
    return this;
  }

  // Appends a 64-bit unsigned integer (little-endian)
  uint64(value) {
    this.ensureCapacity(8);
    this.view.setBigUint64(this.length, BigInt(value), true);
    this.length += 8;
    return this;
  }

  // Appends a 32-bit signed integer (little-endian)
  int32(value) {
    this.ensureCapacity(4);
    this.view.setInt32(this.length, value, true);
    this.length += 4;
    return this;
  }

  bool(value) {
    this.ensureCapacity(1);
    this.view.setUint8(this.length, value ? 1 : 0);
    this.length += 1;
    return this;
  }

  // Appends a string (length-prefixed with 32-bit unsigned integer)
  string(value) {
    const encoder = new TextEncoder();
    const bytes = encoder.encode(value);
    this.ensureCapacity(4 + bytes.length);
    this.view.setUint32(this.length, bytes.length, true);
    this.length += 4;
    new Uint8Array(this.buffer, this.length, bytes.length).set(bytes);
    this.length += bytes.length;
    return this;
  }

  // Ensures the buffer has enough capacity, resizing if needed
  ensureCapacity(additional) {
    const required = this.length + additional;
    if (required > this.buffer.byteLength) {
      const newCapacity = Math.max(required, this.buffer.byteLength * 2 || 16);
      const newBuffer = new ArrayBuffer(newCapacity);
      new Uint8Array(newBuffer).set(new Uint8Array(this.buffer));
      this.buffer = newBuffer;
      this.view = new DataView(this.buffer);
    }
  }
}
